src/java.base/share/classes/java/util/zip/ZipInputStream.java

```diff
@@ -578,7 +578,8 @@ private void readEnd(ZipEntry e) throws IOException {
         }
         if ((flag & 8) == 8) {
             /* "Data Descriptor" present */
-            if (inf.getBytesWritten() > ZIP64_MAGICVAL ||
+            if (hasZip64Extra(e) ||
```

Comment:
I think it would be better to change the ordering of this boolean conditions. Instead of the current proposed:

```java
if (hasZip64Extra(e) ||
    inf.getBytesWritten() > ZIP64_MAGICVAL ||
    inf.getBytesRead() > ZIP64_MAGICVAL) {

```
the new condition being introduced here - the `hasZip64Extra(e)` should be at the end:

```java
if (inf.getBytesWritten() > ZIP64_MAGICVAL ||
    inf.getBytesRead() > ZIP64_MAGICVAL ||
    hasZip64Extra(e)) {

```

That way the chances of this change running additional code (and thus any unexpected semantic change) for existing zip/jar entries which the Inflater already considers as zip64 is avoided.